### PR TITLE
Add concurrent operation for uploading and cancellation of background tasks

### PR DESCRIPTION
**Pull request by travis79**
I tried to break this down into logical commits for easier review.  I would recommend reviewing commit by commit, if possible.

### CODE REVIEW

File: glean-core/ios/Glean/Dispatchers.swift [hunk 1]
```diff
@@ -88,17 +84,66 @@ class Dispatchers {
...
+            // If we are in testing mode, go ahead and wait until it is finished before continuing
+            // to ensure synchronous execution of the task.
             if testingMode {
-                // If we are in testing mode, go ahead and execute the operation and wait until it
-                // is finished before continuing to give us synchronous execution of the task.
-                operation.start()
-                operation.waitUntilFinished()
-            } else {
-                operationQueue.addOperation(operation)
+                serialOperationQueue.waitUntilAllOperationsAreFinished()
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: That's a mouth-full.

File: glean-core/ios/Glean/Scheduler/GleanOperation.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// Base class for other Glean Operations that contains the management of
+/// the executing and finished key-value coding/observers
+class GleanOperation: Operation {
+
+    private var _executing = false {
```
Comment (badboy) [inline glean-core/ios/Glean/Scheduler/GleanOperation.swift]: Can you explain this block of code? It looks ... weird

File: glean-core/ios/Glean/Glean.swift [hunk 3]
```diff
@@ -111,6 +111,11 @@ public class Glean {
         if isInitialized() {
             glean_set_upload_enabled(handle, enabled.toByte())
         }
+
+        // Cancel background metric recording and upload tasks
+        if !uploadEnabled {
```
Comment (badboy) [inline glean-core/ios/Glean/Glean.swift]: In #288 this function is changed. We probably should land that first, then change it here to cancel background tasks before we disable stuff.

File: glean-core/ios/GleanTests/DispatchersTest.swift [hunk 4]
```diff
@@ -106,4 +106,76 @@ class DispatchersTest: XCTestCase {
...
+        // Now cancel the tasks
+        Dispatchers.shared.cancelBackgroundTasks()
+
+        // Assert the task was removed from the queue
+        XCTAssertEqual(Dispatchers.shared.preInitOperations.count, 0, "Task must be removed")
+
+        // Set up our test conditions for normal execution by setting queuing to false
+        Dispatchers.shared.setTaskQueuing(enabled: false)
+
+        // Create some counters to use to determine if the tasks were actually cancelled
```
Comment (badboy) [inline glean-core/ios/GleanTests/DispatchersTest.swift]: Can you split this into 2 tests? First testing that we clear preInit operations, second that we clear others.

(Maybe even split into preInit/sync/async?)

File: glean-core/ios/GleanTests/DispatchersTest.swift [hunk 5]
```diff
@@ -106,4 +106,76 @@ class DispatchersTest: XCTestCase {
...
+        let concurrentOperation = BlockOperation()
+        concurrentOperation.addExecutionBlock {
+            while !concurrentOperation.isCancelled {
+                asyncCounter += 1
+            }
+        }
+        Dispatchers.shared.launchAsync(operation: concurrentOperation)
+
+        // Let the tasks run for 1 second
+        sleep(1)
```
Comment (badboy) [inline glean-core/ios/GleanTests/DispatchersTest.swift]: Hmmm, not a fan, but I guess this is what we have to do.

File: glean-core/ios/Glean/Dispatchers.swift [hunk 6]
```diff
@@ -88,17 +84,66 @@ class Dispatchers {
...
+    /// This function is used to execute tasks in an asynchrounous manner and still give us the ability
+    /// to cancel the tasks by creating them as `Operation`s rather than using GCD.
+    ///
+    /// This function specifically ignores the `queueInitialTasks` flag because the only tasks that
+    /// should be launched by this are the ping upload schedulers and those should run regardless of
+    /// the initialized state.
+    ///
+    /// If `testingMode` is enabled, then launch will just execute the task rather than adding it to the
+    /// concurrent queue to avoid asynchronous issues while testing
+    func launchAsync(block: @escaping () -> Void) {
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Dispatchers.swift]: Looks like the logic here is the same of the other `launchAsync`. Can we call the other instead of duplicating, here? Looks like the only difference is in the `block`

File: glean-core/ios/Glean/Glean.swift [hunk 7]
```diff
@@ -152,7 +152,10 @@ public class Glean {
     /// If the ping currently contains no content, it will not be assembled and
     /// queued for sending.
     func sendPingsByName(pingNames: [String]) {
-        _ = Dispatchers.shared.launch {
+        // This runs in the background to allow for processing of pings off of the
+        // main thread. Please note that the ping uploader will spawn other async
+        // operations if there are pings to upload.
+        Dispatchers.shared.launchAsync {
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Glean.swift]: nit: maybe this could be just `launch`?

File: glean-core/ios/Glean/Scheduler/GleanOperation.swift (hunk 2 above)
Comment (Dexterp37) [reply glean-core/ios/Glean/Scheduler/GleanOperation.swift]: I second this

File: glean-core/ios/Glean/Scheduler/PingUploadOperation.swift [hunk 8]
```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+class PingUploadOperation: GleanOperation {
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Scheduler/PingUploadOperation.swift]: nit: can you please add docs for this?

File: glean-core/ios/GleanTests/DispatchersTest.swift [hunk 9]
```diff
@@ -106,4 +106,76 @@ class DispatchersTest: XCTestCase {
...
+
+        let concurrentOperation = BlockOperation()
+        concurrentOperation.addExecutionBlock {
+            while !concurrentOperation.isCancelled {
+                asyncCounter += 1
+            }
+        }
+        Dispatchers.shared.launchAsync(operation: concurrentOperation)
+
+        // Let the tasks run for 1 second
```
Comment (Dexterp37) [inline glean-core/ios/GleanTests/DispatchersTest.swift]: Why is this needed? Can you document the rationale?

File: glean-core/ios/Glean/Dispatchers.swift [hunk 10]
```diff
@@ -20,21 +20,29 @@ class Dispatchers {
...
         queue.maxConcurrentOperationCount = 1
         return queue
     }()
 
+    // This is a task queue for all Glean background operations that need be executed concurrently.
+    // It is currently set to use a maximum of 4 concurrent operations at a time.
+    lazy var concurrentOperationsQueue: OperationQueue = {
+        var queue = OperationQueue()
+        queue.name = "Glean concurrent dispatch queue"
+        queue.maxConcurrentOperationCount = 4
```
[the remark this answers is not in the document]
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: That's fine, it was an arbitrary choice of number and I didn't realize that the default was determined by the OS.

File: glean-core/ios/Glean/Dispatchers.swift [hunk 11]
```diff
@@ -43,39 +51,27 @@ class Dispatchers {
...
-
+    /// recording of metrics and things that need to execute in order. Since this executes the tasks on
+    /// a non-concurrent (serial) queue, the tasks are executed in the order that they are launched.
+    ///
+    /// If `queueInitialTasks` is enabled, then the operation will be created and added to the
+    /// `preInitOperations` array but not executed until flushed.
+    ///
+    /// If `testingMode` is enabled, then launch will await the execution of the task (unless queuing is
+    /// enabled)
+    func launch(block: @escaping () -> Void) {
```
[the remark this answers is not in the document]
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: It's async (as in it happens on a background thread), but the execution of that particular thread queue is serial rather than concurrent so only one task launched by this function can be executed at a time.  I think the naming of the `launchAsync` functions I added is probably the problem here, they should be `launchConcurrent`.  I intended to have something that we could use that would be like running a task in a coroutine.

File: glean-core/ios/Glean/Dispatchers.swift (hunk 1 above)
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: Yeah, Swift sometimes likes to read like a sentence.

File: glean-core/ios/Glean/Dispatchers.swift (hunk 6 above)
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: I think we can do that, there is no way to convert an operation into a closure but we can do the other way around, so this function can call the one that accepts an operation.

File: glean-core/ios/Glean/Dispatchers.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (Dexterp37) [reply glean-core/ios/Glean/Dispatchers.swift]: Ok, I thought so. I think we should really hint that this is used for the API. We don't want other stuff going on this queue. I'd also drop the async suffix, as everything is async here

File: glean-core/ios/Glean/Glean.swift (hunk 3 above)
Comment (travis79) [reply glean-core/ios/Glean/Glean.swift]: I can leave this out for now

File: glean-core/ios/Glean/Glean.swift (hunk 7 above)
Comment (travis79) [reply glean-core/ios/Glean/Glean.swift]: The bug was to move this off of the serial queue.  If we call this with `launch` and it takes some time to perform the network upload, then the serial execution queue is blocked.

Comment (travis79) [reply glean-core/ios/Glean/Glean.swift]: The `launchAsync` puts this on a different execution queue that is concurrently executed so as to not block the execution of other tasks.

File: glean-core/ios/Glean/Scheduler/GleanOperation.swift (hunk 2 above)
Comment (travis79) [reply glean-core/ios/Glean/Scheduler/GleanOperation.swift]: This is something I just learned about `Operation`.  The `isExecuting` and `isFinished` and other member properties use [KVC](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/KeyValueCoding/index.html) (key-value coding) which provides indirect access to properties and [KVO](https://developer.apple.com/documentation/swift/cocoa_design_patterns/using_key-value_observing_in_swift) (key-value observing) to allow for event driven observation of these properties. 

This is how that signalling looks to trigger events that are hooked into this. A more in-depth explanation can be found in the docs for [`Operation`](https://developer.apple.com/documentation/foundation/operation)

File: glean-core/ios/GleanTests/DispatchersTest.swift (hunk 9 above)
Comment (travis79) [reply glean-core/ios/GleanTests/DispatchersTest.swift]: It's needed just so that the background tasks have some time to run.  I can us uSleep for a shorter delay or attempt to find some other way to guarantee that the tasks are running.

File: glean-core/ios/GleanTests/DispatchersTest.swift (hunk 5 above)
Comment (travis79) [reply glean-core/ios/GleanTests/DispatchersTest.swift]: I don't like sleeps either, but in this case it's just a delay and not waiting for something to occur.  Since both you and @Dexterp37 had issue with this line, let me see if I can find another way to accomplish this.

File: glean-core/ios/GleanTests/DispatchersTest.swift (hunk 9 above)
Comment (Dexterp37) [reply glean-core/ios/GleanTests/DispatchersTest.swift]: No no, that's fine, just extend the comment to mention that.

File: glean-core/ios/Glean/Dispatchers.swift (hunk 11 above)
[the remark this answers is not in the document]
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: I'll rename these to be more descriptive.  `launchAPI` for the serial queue, and I can rename the other to `launchConcurrent` or similar.

File: glean-core/ios/Glean/Scheduler/GleanOperation.swift (hunk 2 above)
Comment (travis79) [reply glean-core/ios/Glean/Scheduler/GleanOperation.swift]: I added some comments to help clarify this

File: glean-core/ios/Glean/Dispatchers.swift [hunk 12]
```diff
@@ -20,21 +20,28 @@ class Dispatchers {
...
+    lazy var serialOperationQueue: OperationQueue = {
         var queue = OperationQueue()
-        queue.name = "Glean dispatch queue"
+        queue.name = "Glean serial dispatch queue"
         queue.maxConcurrentOperationCount = 1
         return queue
     }()
 
+    // This is a task queue for all Glean background operations that need be executed concurrently.
+    // It is currently set to use a maximum of 4 concurrent operations at a time.
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Dispatchers.swift]: nit: this is not set to 4 anymore, but you can see that the OS determines how many concurrent operations :)

File: glean-core/ios/Glean/Dispatchers.swift [hunk 13]
```diff
@@ -88,17 +83,62 @@ class Dispatchers {
...
+    /// concurrent queue to avoid asynchronous issues while testing
+    func launchConcurrent(block: @escaping () -> Void) {
+        launchConcurrent(operation: BlockOperation(block: block))
+    }
 
-        return operation
+    /// Cancel any pending background tasks
+    func cancelBackgroundTasks() {
+        serialOperationQueue.cancelAllOperations()
+        concurrentOperationsQueue.cancelAllOperations()
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Dispatchers.swift]: Note: this will also cancel `sendPingsByName`. Could this result in partial pings or corrupted data?

File: glean-core/ios/Glean/Dispatchers.swift (hunk 12 above)
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: Good catch, thanks!

File: glean-core/ios/Glean/Dispatchers.swift (hunk 13 above)
Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: Since this cancels the network request, I would expect this to not perform a partial upload.  Looking at the [docs](https://developer.apple.com/documentation/foundation/urlsessiontask/1411591-cancel) it looks like it notifies the task so that it can cancel gracefully.

Comment (Dexterp37) [reply glean-core/ios/Glean/Dispatchers.swift]: This is doing ping persistence to disk in addition to the upload. If we cancel (kill the request?) while writing to disk then we result in partial pings being written. Upload doesn't concern me too much

Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: Ahh I see what you are asking.  The answer is, I'm pretty sure it's safe, and here is why:  The way that the `sendPingsByName` function launches on the concurrent queue is by passing a closure block to be executed.  Since this block doesn't do anything to handle the cancel state, it would run to completion.  Cancel in this case does not mean that it is aborting the thread, it means that it signals the operation to cancel.  See the test I added and how I used the isCancelled in the small counter operations I used.  If I didn't do that, then they would run forever (As I found out the first time I tried the test without it).

Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: Want me try to write a test around this?

Comment (Dexterp37) [reply glean-core/ios/Glean/Dispatchers.swift]: Ah, I see now. That makes sense, thanks for explaining. Can you add a small comment above the `concurrentOperationsQueue.cancelAllOperations()` line to explain why this is safe (basically, what you just told me). Writing a test for this would be a bit tricky, I'd say not a blocker for now, but file a bug for it if you feel like :)

Comment (travis79) [reply glean-core/ios/Glean/Dispatchers.swift]: I'll add the comment, and I filed a bug to follow up on making the operation that is calling the function that serializes pings obey the cancel state gracefully [here](https://bugzilla.mozilla.org/show_bug.cgi?id=1582464).

File: glean-core/ios/Glean/Dispatchers.swift [hunk 14]
```diff
@@ -20,21 +20,27 @@ class Dispatchers {
...
+    // It is currently set to be a serial queue by setting the `maxConcurrentOperationsCount` to 1
+    lazy var serialOperationQueue: OperationQueue = {
         var queue = OperationQueue()
-        queue.name = "Glean dispatch queue"
+        queue.name = "Glean serial dispatch queue"
         queue.maxConcurrentOperationCount = 1
         return queue
     }()
 
+    // This is a task queue for all Glean background operations that need be executed concurrently.
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: Actually, we don't guarantee concurrency, it's just a possibility.

```suggestion
    // This is a task queue for all Glean background operations that can be executed concurrently.
```

File: glean-core/ios/Glean/Dispatchers.swift [hunk 15]
```diff
@@ -43,39 +49,27 @@ class Dispatchers {
     // property
     private init() {}
 
-    /// This function launches a background `Operation`
+    /// This function launches a background `Operation` on a serially executed queue.
     ///
     /// - parameters:
     ///   * block: The block of code to execute, as a Closure that accepts no arugments and returns Void
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: Now that I re-read this I think it's weird that we explain the closure. That's what the type system is for.

File: glean-core/ios/Glean/Dispatchers.swift [hunk 16]
```diff
@@ -43,39 +49,27 @@ class Dispatchers {
...
-    func launch(block: @escaping () -> Void) -> Operation {
-        let operation = GleanOperation(block)
-
+    /// recording of metrics and things that need to execute in order. Since this executes the tasks on
+    /// a non-concurrent (serial) queue, the tasks are executed in the order that they are launched.
+    ///
+    /// If `queueInitialTasks` is enabled, then the operation will be created and added to the
+    /// `preInitOperations` array but not executed until flushed.
+    ///
+    /// If `testingMode` is enabled, then launch will await the execution of the task (unless queuing is
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]:
```suggestion
    /// If `testingMode` is enabled, then `launchAPI` will await the execution of the task (unless queuing is
```

File: glean-core/ios/Glean/Dispatchers.swift [hunk 17]
```diff
@@ -88,17 +82,72 @@ class Dispatchers {
...
+    ///   * operation: The `Operation` to execute
+    ///
+    /// This function is used to execute tasks in an asynchrounous manner and still give us the ability
+    /// to cancel the tasks by creating them as `Operation`s rather than using GCD.
+    ///
+    /// This function specifically ignores the `queueInitialTasks` flag because the only tasks that
+    /// should be launched by this are the ping upload schedulers and those should run regardless of
+    /// the initialized state.
+    ///
+    /// If `testingMode` is enabled, then launch will just execute the task rather than adding it to the
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]:
```suggestion
    /// If `testingMode` is enabled, then `launchConcurrent` will just execute the task rather than adding it to the
```

File: glean-core/ios/Glean/Dispatchers.swift [hunk 18]
```diff
@@ -88,17 +82,72 @@ class Dispatchers {
...
-                operation.start()
-                operation.waitUntilFinished()
-            } else {
-                operationQueue.addOperation(operation)
+                serialOperationQueue.waitUntilAllOperationsAreFinished()
             }
         }
+    }
+
+    /// This function launches a background `Operation` on a concurrently executed queue.
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: The operation isn't "background". IMO leaving it out is just as clear.

```suggestion
    /// This function launches an `Operation` on a concurrently executed queue.
```

File: glean-core/ios/Glean/Dispatchers.swift [hunk 19]
```diff
@@ -88,17 +82,72 @@ class Dispatchers {
...
+        if testingMode {
+            operation.start()
+            operation.waitUntilFinished()
+        } else {
+            concurrentOperationsQueue.addOperation(operation)
+        }
+    }
 
-        return operation
+    /// This function launches a background block of code on a concurrently executed queue.
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]: The blcok isn't "background". IMO leaving it out is just as clear.
```suggestion
    /// This function launches a block of code on a concurrently executed queue.
```

File: glean-core/ios/Glean/Dispatchers.swift [hunk 20]
```diff
@@ -88,17 +82,72 @@ class Dispatchers {
...
+        // This will remove all queued operations prior to Glean.initialize() so that they won't be
+        // executed if flushQueuedInitialTasks() is called
+        preInitOperations.removeAll()
+
+        // This will cancel operations in the serially executed queue. This includes most of the
+        // metrics recording and things that need to execute in order. This would not stop the currently
+        // executing operation, but would prevent all remaining operations from executing.
+        serialOperationQueue.cancelAllOperations()
+
+        // This will cancel operations that are executing concurrently. This doesn't abort the thread
```
Comment (badboy) [inline glean-core/ios/Glean/Dispatchers.swift]:
```suggestion
        // This will cancel operations that are executing concurrently. 
        // This doesn't abort running operations immediately
```

File: glean-core/ios/Glean/Metrics/StringMetric.swift [hunk 21]
```diff
@@ -47,7 +47,7 @@ public class StringMetricType {
     public func set(_ value: String) {
         guard !self.disabled else { return }
 
-        _ = Dispatchers.shared.launch {
+        _ = Dispatchers.shared.launchAPI {
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/StringMetric.swift]: We don't need the assignment anymore, do we?

File: glean-core/ios/Glean/Metrics/TimespanMetric.swift [hunk 22]
```diff
@@ -56,7 +56,7 @@ public class TimespanMetricType {
 
         let startTime = timestampNanos()
 
-        _ = Dispatchers.shared.launch {
+        _ = Dispatchers.shared.launchAPI {
```
Comment (badboy) [inline glean-core/ios/Glean/Metrics/TimespanMetric.swift]: As above and below a couple of times. no `_ =` needed

Comment (travis79) [reply glean-core/ios/Glean/Metrics/TimespanMetric.swift]: There were several places this still existed, I squashed them all.
